### PR TITLE
Fixed failure in getBasePointer of HostBufferIntern

### PR DESCRIPTION
**Pull request by erikzenker**
- Uses this->getPointer as much as possible
- Fixes getBasePointer
- Renames ownPointer to ownsPointer
- [ ] merge after #1283

### CODE REVIEW

File: src/libPMacc/include/memory/buffers/HostBufferIntern.hpp [hunk 1]
```diff
@@ -70,9 +76,9 @@ class HostBufferIntern : public HostBuffer<TYPE, DIM>
     {
         __startOperation(ITask::TASK_HOST);
 
-        if (pointer && ownPointer)
+        if (this->getPointer() && ownsPointer)
```
Comment (psychocoderHPC) [inline src/libPMacc/include/memory/buffers/HostBufferIntern.hpp]: This should be `this->getBasePointer()`

File: src/libPMacc/include/memory/buffers/HostBufferIntern.hpp [hunk 2]
```diff
@@ -70,9 +76,9 @@ class HostBufferIntern : public HostBuffer<TYPE, DIM>
...
         {
-            CUDA_CHECK(cudaFreeHost(pointer));
+            CUDA_CHECK(cudaFreeHost(this->getPointer()));
```
Comment (psychocoderHPC) [inline src/libPMacc/include/memory/buffers/HostBufferIntern.hpp]: Should be `CUDA_CHECK(cudaFreeHost(this->getBasePointer()));`

File: src/libPMacc/include/memory/buffers/HostBufferIntern.hpp [hunk 3]
```diff
@@ -70,9 +76,9 @@ class HostBufferIntern : public HostBuffer<TYPE, DIM>
...
         }
```
Comment (psychocoderHPC) [inline src/libPMacc/include/memory/buffers/HostBufferIntern.hpp]: Please set both pointer `pointer` and `basePointer` to `NULL` after the if condition, this helps to detect if already freed memory is used.

**Comment by psychocoderHPC** [comment]
I add the flag **feature** and not **bug** to this pull request. The reason is that this was no error only a not optimal implementation.

**Comment by erikzenker** [comment]
If the PR gets merged, then there might be problems with methods calling getBasePointer from HostBufferIntern. Because it might be possible that they wanted to call getPointer. 

I found one example for such a behaviour in HostBuffer when [cartBuffer()](https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/src/libPMacc/include/memory/buffers/HostBuffer.hpp#L72-L86) is called. I think the implementer does not want to get the base pointer there.

File: src/libPMacc/include/memory/buffers/HostBufferIntern.hpp (hunk 1 above)
Comment (erikzenker) [inline src/libPMacc/include/memory/buffers/HostBufferIntern.hpp]: fixed

File: src/libPMacc/include/memory/buffers/HostBufferIntern.hpp (hunk 2 above)
Comment (erikzenker) [inline src/libPMacc/include/memory/buffers/HostBufferIntern.hpp]: fixed

File: src/libPMacc/include/memory/buffers/HostBufferIntern.hpp (hunk 3 above)
Comment (erikzenker) [inline src/libPMacc/include/memory/buffers/HostBufferIntern.hpp]: fixed

**Comment by psychocoderHPC** [comment]
We must fix the CuSTL parts within this pull request, else we creates a not working version of PMacc.

**Comment by psychocoderHPC** [comment]
@Heikman please take a look which pointer is needed [here](https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/src/libPMacc/include/memory/buffers/HostBuffer.hpp#L72-L86)

**Comment by psychocoderHPC** [comment]
I opened an new issue for the wrong usage of `getPointer()` and the corresponding `getPitched()` for device buffer.
We should first solve #1283 and than merge this one.

**Comment by Heikman** [comment]
cuSTL/HostBuffer needs the BasePointer because it owns it, therefore it releases it at destruction.

**Comment by psychocoderHPC** [comment]
we should switch to #1283 for the `getPointer()` discussion

**Comment by sbastrakov** [comment]
I believe this can be closed.

**Comment by psychocoderHPC** [comment]
I think this issue is still not solved. But would also not start to fix it because @michaelsippel is refactoring this part anyway.

**Pull request closed without merging**